src/storage/__tests__/sqlite-test.js

```diff
@@ -136,4 +140,143 @@ describe('expo-sqlite', () => {
     // Instead, we get:
     expect(data).toEqual({ double: 8 }); // bad: missing the second INSERT
   });
+
+  describe('transactions failing in app code', () => {
```

Comment:
```
expo-sqlite tests: Show transactions' (non-)handling of exceptions
```

Is there a connection between the "(non-)handling of exceptions" and the "broken asynchrony handling" (from an earlier commit) that you'd like to highlight? For example, is one a consequence of the other in some interesting way?

Or can I think of these as separate messes that we're cleaning up, as I try to follow everything we're doing in this PR? 🙂

Reply:
I think they can basically be thought of as separate.

In particular, the Web standards approached these issues in basically the opposite order:
* The Web SQL Database spec [requires](https://www.w3.org/TR/webdatabase/#processing-model) that exceptions be handled, in a basically reasonable way: if the transaction callback, or a statement callback, raises, then the transaction gets rolled back.
  * So the fact that node-websql doesn't do this is simply a bug.
  * (Reading closely, in the case of a statement error callback the text isn't totally clear about what happens if it raises. But I believe it does end up requiring that that's a rollback -- "the error callback did not return false" includes the case where the error callback didn't return anything and instead threw.)
* But the eager auto-committing that's incompatible with asynchrony in the app code is faithful to the spec.

  And the IndexedDB spec, which is the actual Web standard that came after it, makes the same choice and [explicitly explains it](https://w3c.github.io/IndexedDB/#ref-for-transaction-commit%E2%91%A2):
  > Transactions are expected to be short lived. This is encouraged by the automatic committing functionality described below.
  > > Authors can still cause transactions to run for a long time; however, this usage pattern is not advised as it can lead to a poor user experience.

So in a perfect implementation of the Web SQL draft standard, or of the related actual Web standard, the second of our issues is fixed, but the one we fix first remains unfixed.

---

Re that warning in the IndexedDB spec:

I think what they have in mind for "Authors can still cause transactions to run for a long time" is basically to do something just like our workaround, and to do so in a situation where the thing the workaround waits for is something that potentially takes a truly long time, like a network request. Then the user's device would sit there with its CPU spinning in a loop of `SELECT 1` filler requests (just spelled differently because IndexedDB isn't SQL), while the request takes 60 seconds timing out.

We'll need to be mindful not to put real slow non-database steps in the middle of a transaction, which would be a similarly bad user experience. But I have more confidence in us avoiding that than the browser developers have (reasonably!) in the authors of, say, the ad-targeting scripts that are all over any news site and much of the rest of the Web.

Reply:
Great, thanks! This is very helpful.